proposals/lip-0039.md

```diff
@@ -302,273 +328,365 @@ proof={siblingHashes:[cc956a85, 3c516152, e041e1c0, 6400721e, 3c32b131],
+            high = middle
         else:
-            prepend 1 to queryBitmap
-            append currentNodeSibling.hash to siblingHashes
-        set h to h+1
-
-    // currentNode is an empty default node or a leaf node
-    if currentNode is defaultEmptyNode:
-        // queryKey is not part of the tree
-        let response be a data structure with:
-            {key=queryKey,
-            value=empty buffer,
-            binaryBitmap=queryBitmap,
-            siblingHashes=siblingHashes,
-            ancestorHashes=ancestorHashes}
-    else:
-        let response be a data structure with:
-            {key=currentNode.key,
-            value=currentNode.value,
-            binaryBitmap=queryBitmap,
-            siblingHashes=siblingHashes,
-            ancestorHashes=ancestorHashes}
-    return response
+            low = middle
+
+    insert_index = high
+
+    if insert_index == len(queries):
+        return queries + [q]
+
+    p = queries[insert_index]
+    if p.binary_path != q.binary_path: # If another query at the same 
+        queries.insert(insert_index, q)
+
+    return queries
 ```
 
-Note that the first digit of `queryBitmap` is always a 1, else the tree would have been invalidly constructed. As a consequence, the hex encoded value of `queryBitmap` corresponds to a unique series of digits.
-To generate a complete inclusion proof, several query responses are combined together, and their sibling hashes are merged in a single array. We first define the utility function `sortByBitmapAndKey`.
+The following function generates the query response to a single key, including the sibling hashes and the hash of the visited nodes.
 
-```java
-function sortByBitmapAndKey(queries):
-    sort queries by the longest binaryBitmap, breaking ties by smaller key.
-    // this corresponds to a bottom-up left-right ordering of the corresponding leaf nodes
+```python
+def generate_query_proof(query_key: bytes, root: bytes) -> QueryProof:
+    ancestor_hashes = []
+    sibling_hashes = []
+    current_node = get_node(root)
+    binary_bitmap = ""
+
+    # Starting from the root, we navigate down the tree.
+    h = 0
+    while isinstance(current_node, BranchNode):
+        ancestor_hashes.append(current_node.hash)
+        if is_bit_set(query_key, h):
+            sibling_hash = current_node.left_hash
+            current_node = get_node(current_node.right_hash)
+        else:
+            sibling_hash = current_node.right_hash
+            current_node = get_node(current_node.left_hash)
+
+        if sibling_hash == EmptyNode.hash:
+            binary_bitmap = "0" + binary_bitmap
+        else:
+            binary_bitmap = "1" + binary_bitmap
+            sibling_hashes.append(sibling_hash)
+        h += 1
+
+    if isinstance(current_node, EmptyNode):
+        # query_key is not part of the tree
+        return QueryProof(
+            query_key, EMPTY_VALUE, binary_to_bytes(binary_bitmap), ancestor_hashes, sibling_hashes
```

Comment:
I think the function `binary_to_bytes` was not introduced.

Reply:
I just calculate the conversion explicitly